src/lib.rs

```diff
@@ -12,61 +14,130 @@ pub struct LinkDrop {
+        assert_eq!(
+            env::predecessor_account_id(),
+            env::current_account_id(),
+            "Claim only can come from this account"
+        );
         let amount = self
             .accounts
             .remove(&env::signer_account_pk())
             .expect("Unexpected public key");
         Promise::new(env::current_account_id()).delete_key(env::signer_account_pk());
-        Promise::new(env::predecessor_account_id()).transfer(amount)
+        Promise::new(account_id).transfer(amount)
     }
 
     /// Create new account and and claim tokens to it.
-    pub fn create_account_and_claim(&mut self, new_account_id: AccountId, new_public_key: String) -> Promise {
-        assert_eq!(env::signer_account_id(), env::current_account_id());
+    pub fn create_account_and_claim(
+        &mut self,
+        new_account_id: AccountId,
+        new_public_key: Base58PublicKey,
+    ) -> Promise {
+        assert_eq!(
+            env::predecessor_account_id(),
+            env::current_account_id(),
+            "Create account and claim only can come from this account"
+        );
+        assert!(
+            env::is_valid_account_id(new_account_id.as_bytes()),
+            "Invalid account id"
+        );
         let amount = self
             .accounts
-            .remove(&env::signer_account_pk())
+            .get(&env::signer_account_pk())
             .expect("Unexpected public key");
-        Promise::new(env::current_account_id()).delete_key(env::signer_account_pk());
         Promise::new(new_account_id)
             .create_account()
-            .add_full_access_key(str_to_public_key(new_public_key))
+            .add_full_access_key(new_public_key.into())
             .transfer(amount)
+            .then(ext_self::on_account_created_and_claimed(
+                &env::current_account_id(),
+                NO_DEPOSIT,
+                ON_CREATE_ACCOUNT_CALLBACK_GAS,
+            ))
+    }
+
+    /// Create new account without linkdrop and deposit passed funds (used for creating sub accounts directly).
+    #[payable]
+    pub fn create_account(
+        &mut self,
+        new_account_id: AccountId,
+        new_public_key: Base58PublicKey,
+    ) -> Promise {
+        assert!(
+            env::is_valid_account_id(new_account_id.as_bytes()),
+            "Invalid account id"
+        );
+        let amount = env::attached_deposit();
+        Promise::new(new_account_id)
+            .create_account()
+            .add_full_access_key(new_public_key.into())
+            .transfer(amount)
+    }
+
+    /// Called after account creation & claim was called.
+    pub fn on_account_created_and_claimed(&mut self) -> bool {
+        let creation_succeeded = is_promise_success();
+        if creation_succeeded {
+            Promise::new(env::current_account_id()).delete_key(env::signer_account_pk());
+            self.accounts.remove(&env::signer_account_pk());
+        }
```

Comment:
In case the transfer failed, you can put the amount back:
```suggestion
        } else { 
            let pk = env::signer_account_pk(); 
            let value = self.accounts.get(&pk).unwrap_or(0);
            self.accounts.insert(&pk, (value + amount.0).into());
        }
```